src/handler/mod.rs

```diff
@@ -406,6 +472,29 @@ impl Handler {
     ) {
         if request_call.retries() >= self.request_retries {
             trace!("Request timed out with {}", node_address);
+            if let Some(relay) = self
+                .nat_hole_puncher
+                .new_peer_latest_relay
+                .remove(&node_address.node_id)
```

Comment:
it falls on the relay, no point to consume memory. a relay shouldn't pass us any nodes it didn't have in it k-buckets recently by spec. https://github.com/ethereum/devp2p/blob/discv5-v5.2/discv5/discv5-theory.md#message-flow
this implementation only stores relays for targets it can contact https://github.com/emhane/discv5/blob/5b750584f832d44bb2fc8049295797d5fee86a18/src/handler/mod.rs#L1179